// model-storage-driver/build/indexed-db-adapter.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var idb_keyval_1 = require("idb-keyval");
var IndexedDbAdapter = /** @class */ (function () {
    function IndexedDbAdapter() {
    }
    /**
     * init indexedDb storage
     *
     * @returns       any
     */
    IndexedDbAdapter.prototype.init = function (storageName) {
        return __awaiter(this, void 0, void 0, function () {
            var localMemory;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.storageName = storageName;
                        localMemory = [];
                        return [4 /*yield*/, idb_keyval_1.get(this.storageName)
                                .then(function (value) { return localMemory = value ? value : localMemory; })
                                .catch(function () { return _this.memoryCache = []; })];
                    case 1:
                        _a.sent();
                        this.storedData = localMemory;
                        return [4 /*yield*/, this.__persist()];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * set storage value
     *
     * @param cache new value for storedData
     * @returns void
     */
    IndexedDbAdapter.prototype.set = function (cache) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.storedData = cache;
                        return [4 /*yield*/, this.__persist()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * push element to indexedDb storage memory
     *
     * @param el    the pushed element `value`.
     * @returns void
     */
    IndexedDbAdapter.prototype.push = function (el) {
        return __awaiter(this, void 0, void 0, function () {
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this;
                        return [4 /*yield*/, this.get()];
                    case 1:
                        _a.storedData = _b.sent();
                        this.storedData.push(el);
                        return [4 /*yield*/, this.__persist()];
                    case 2:
                        _b.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * shift element from the indexedDb storage
     *
     * @returns A promise that is resolved with the element value.
     */
    IndexedDbAdapter.prototype.shift = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, element;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this;
                        return [4 /*yield*/, this.get()];
                    case 1:
                        _a.storedData = _b.sent();
                        element = this.storedData.shift();
                        if (!element) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.__persist()];
                    case 2:
                        _b.sent();
                        _b.label = 3;
                    case 3: return [2 /*return*/, element];
                }
            });
        });
    };
    /**
     * get storage value
     *
     * @returns   MemoryMap Object
     */
    IndexedDbAdapter.prototype.get = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, idb_keyval_1.get(this.storageName)
                    // return this.storedData;
                ];
            });
        });
    };
    /**
     * delete all objects from indexedDb storage
     *
     * @returns       void
     */
    IndexedDbAdapter.prototype.reset = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.storedData = [];
                        return [4 /*yield*/, this.__persist()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * persist indexedDb storage data
     *
     * @returns void
     */
    IndexedDbAdapter.prototype.__persist = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, idb_keyval_1.set(this.storageName, this.storedData).catch(function () {
                            _this.memoryCache[_this.storageName] = _this.storedData;
                        })];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    return IndexedDbAdapter;
}());
exports.IndexedDbAdapter = IndexedDbAdapter;
