// model-storage-driver/build/model.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var CryptoJS = __importStar(require("crypto-js"));
var storage_manager_1 = require("./storage-manager");
var Model = /** @class */ (function () {
    function Model() {
        this._data = {};
        this.initialize();
        this.storage = new storage_manager_1.StorageManager(this.storageType, this.storageName);
    }
    Model.prototype.getAll = function () {
        return __awaiter(this, void 0, void 0, function () {
            var allData;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.storage.get()];
                    case 1:
                        allData = _a.sent();
                        return [2 /*return*/, this.decrypt(allData)];
                }
            });
        });
    };
    Model.prototype.shift = function () {
        return __awaiter(this, void 0, void 0, function () {
            var element;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.storage.shift()];
                    case 1:
                        element = _a.sent();
                        if (element !== undefined) {
                            return [2 /*return*/, this.decrypt(element)];
                        }
                        else {
                            throw new Error('Empty queue');
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    Model.prototype.save = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _i, _a, field, encryptedData;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        for (_i = 0, _a = this.fields; _i < _a.length; _i++) {
                            field = _a[_i];
                            this._data[field] = this[field];
                        }
                        encryptedData = this.encrypt(this._data);
                        return [4 /*yield*/, this.storage.push(encryptedData)];
                    case 1:
                        _b.sent();
                        this.reset();
                        return [2 /*return*/];
                }
            });
        });
    };
    Object.defineProperty(Model.prototype, "data", {
        get: function () {
            var data = {};
            for (var key in this.fields) {
                data[this.fields[key]] = this[this.fields[key]];
            }
            return data;
        },
        enumerable: true,
        configurable: true
    });
    Model.prototype.reset = function () {
        this._data = {};
        for (var _i = 0, _a = this.fields; _i < _a.length; _i++) {
            var key = _a[_i];
            var value = this[key];
            this[key] = (value && value instanceof Object) ? {} : null;
        }
    };
    Model.prototype.encrypt = function (data) {
        var jsonData;
        try {
            jsonData = JSON.stringify(data);
        }
        catch (e) {
            throw new Error('Could not stringify data.');
        }
        return CryptoJS.AES.encrypt(jsonData, 'my-secret-key').toString();
    };
    Model.prototype.decrypt = function (data) {
        var jsonData;
        var output;
        if (Array.isArray(data)) {
            output = data.map(function (v) {
                var bytes = CryptoJS.AES.decrypt(v, 'my-secret-key');
                var originalText = bytes.toString(CryptoJS.enc.Utf8);
                try {
                    jsonData = JSON.parse(originalText);
                }
                catch (e) {
                    throw new Error('Could not parse JSON');
                }
                return jsonData;
            });
        }
        else {
            var bytes = CryptoJS.AES.decrypt(data, 'my-secret-key');
            var originalText = bytes.toString(CryptoJS.enc.Utf8);
            try {
                jsonData = JSON.parse(originalText);
            }
            catch (e) {
                throw new Error('Could not parse JSON');
            }
            output = jsonData;
        }
        return output;
    };
    return Model;
}());
exports.Model = Model;
